content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,216 @@
+
+**Step 6: create your searches method in the controller, its view and routes.**
+
+My controller method looks just like this
+
+```ruby
+ def search
+   return unless query.present?
+
+   @posts = Post.our_first_query(query)
+ end
+```
+
+Also, I did this simple view with a little of Bootstrap.
+
+![Search by semantic information](images/semantic_search.png)
+
+As you can see, I created a bunch of posts using [faker]([https://github.com/faker-ruby/faker](https://github.com/faker-ruby/faker)).
+
+Take a look into that search again… Don’t you see something special? We looked for ‘_darkness_’ and although we don’t have any matching result, we can see that the ones containing ‘_dark_’ were displayed. This is because of the text analyzers. They know these two words are not the same but they do have similar semantics.
+
+Pretty easy, wasn’t it?
+
+What happens if we look for ‘_dakr_’ instead of ‘_dark_’?
+
+![No results found](images/with_typos.png)
+
+Oops, no results. Let’s fix this.
+
+**Step 7: searching with typos and highlighting the match in searches’ results.**
+
+Our search will now look like this:
+
+```ruby
+def self.highlight(query)
+ search(
+   query: {
+     multi_match: {
+       query: query,
+       fuzziness: 'AUTO',
+       fields: %w[title body]
+     }
+   },
+   highlight: {
+     pre_tags: ['<b>'],
+     post_tags: ['</b>'],
+     fields: {
+       body: {}
+     }
+   }
+ )
+end
+```
+
+In this case we are basically adding two new features: _fuzziness_ and _highlight_.
+
+With `fuzziness` we are making Elastic know that errors may appear in the query. This way, it will try to match similar words, so we now will not only cover semantic but syntactic errors too.
+
+With `highlight` we are making the match to be highlighted by adding pre and post tags (in this case HTML bold, but you can add whatever you want).
+
+Let’s now search for ‘_dakrness_’: a word that is neither in our database nor is well written.
+
+![Final results query](images/final_result.png)
+
+## Conclusions
+
+Although in this post we’ve just covered the tip of the iceberg, we can see the potential of this search engine and how easy it is to integrate it to a Rails application.
+
+Elastic also has several other applications. For example, we have been using it in some of our applications not only to admit typos and make searches smarter and faster but also to add [synonyms to some list of words]([https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html)) in our e-commerce apps where we expect searches like ‘ps4’ to match with ‘playstation 4’, ‘playstation four’, or ‘ipad’ to match ‘i pad’ or ‘i-pad’. It also has several interesting uses in [geospacial queries](https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-queries.html) including geodistances and geoshapes.
+
```

Comment:
This part doesn't look like a conclusion. It's like an "additional info" section.
For example, you start saying: `Elastic also has several other applications. For example, w...`
I would like something like "Elastic is a great tool to XXX and YYY, and also ZZZ"

Reply:
Yep, @missx told me something similar. Maybe I have to work a little bit more on this.